opt/memo: use virtual column stats for matching scalar expressions

michae2 opened the pull request:
As of #120668 we now use statistics on virtual computed columns in
statistics builder. Simple queries that synthesize virtual columns in
Project expressions already benefit, because they use the virtual column
ID when synthesizing the virtual column.

Other expressions, however, do not directly use the virtual column ID
when synthesizing a virtual column. This includes Select expressions,
joins, constrained scans, and some Project expressions.

For example, consider a query like the following:

```
CREATE TABLE ab (a INT PRIMARY KEY, b INT AS (a % 10) VIRTUAL, INDEX (b));
SELECT * FROM ab WHERE a % 10 > 3;
```

Even though the filter condition is in terms of `a`, we'd like to use
the statistics on virtual computed column `b` since the expression
matches.

In order to do this, we replace `a % 10` with `b` in a copy of the
filter condition before doing any stats calculations. Then we perform
our normal stats calculations, using `b`.

Fixes: #68254
Fixes: #110146

Epic: CRDB-8949

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/120875)
<!-- Reviewable:end -->

michae2 committed:
opt/memo: use virtual column stats for matching scalar expressions

As of #120668 we now use statistics on virtual computed columns in
statistics builder. Simple queries that synthesize virtual columns in
Project expressions already benefit, because they use the virtual column
ID when synthesizing the virtual column.

Other expressions, however, do not directly use the virtual column ID
when synthesizing a virtual column. This includes Select expressions,
joins, constrained scans, and some Project expressions.

For example, consider a query like the following:

```
CREATE TABLE ab (a INT PRIMARY KEY, b INT AS (a % 10) VIRTUAL, INDEX (b));
SELECT * FROM ab WHERE a % 10 > 3;
```

Even though the filter condition is in terms of `a`, we'd like to use
the statistics on virtual computed column `b` since the expression
matches.

In order to do this, we replace `a % 10` with `b` in a copy of the
filter condition before doing any stats calculations. Then we perform
our normal stats calculations, using `b`.

Fixes: #68254
Fixes: #110146

Epic: CRDB-8949

Release note: None

DrewKimball reviewed (approved):
I had a nit, and there are a few cases where I think this won't work. But the current state <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> if we're covering the cases we care about. Great work!

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120875#-:-NteWOVN9i2Ug0_BMvuz:b-v561el)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4995 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120875#-NteOVFf8Ps4JfnQrpo-:-NteOVFf8Ps4JfnQrpo0:b-dszfa8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7531502631fe5f07575492318187aca8365ab53/pkg/sql/opt/memo/statistics_builder.go#L4995)):*
> ```Go
> 	// and also prune any virtual cols whose outer columns are not contained in
> 	// the expression's outer columns.
> 	type virtExpr struct {
> ```

[nit] There's no concurrency, so you could keep a scratch slice on the `statisticsBuilder`
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 5023 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120875#-NteT9myDNZxUOa2bRu3:-NtePDvK2D_IqLaVloCK:b-qu3cja) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7531502631fe5f07575492318187aca8365ab53/pkg/sql/opt/memo/statistics_builder.go#L5023)):*
> ```Go
> 	replace = func(e opt.Expr) opt.Expr {
> 		for _, ve := range virtExprs {
> 			if e == ve.expr {
> ```

Note: I believe this is relying on the expression de-duplication performed in the optbuilder to work when the user explicitly writes the expression:
```
https://github.com/cockroachdb/cockroach/blob/4a9385cacb82e7a8d6d37e5d9a26a6b7c845aab6/pkg/sql/opt/optbuilder/project.go#L278-L285
```
I think it wouldn't work if you pruned the `b` column before the filter:
```
SELECT * FROM (SELECT a FROM abc) WHERE a % 10 > 3
```
It should also fail if the expression was copied at any point, which could motivate testing with prepared statements and routines. Prepared statements in particular seem like they could be a problematic omission, since they're so common.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Thanks for looking at it!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120875#-:-Nts8U7L8giwx7xnvHdh:b-ers4pd)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4995 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120875#-NteOVFf8Ps4JfnQrpo-:-Nts8WG5Eq06XDtbYwBQ:b-rjjiyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7531502631fe5f07575492318187aca8365ab53/pkg/sql/opt/memo/statistics_builder.go#L4995)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] There's no concurrency, so you could keep a scratch slice on the `statisticsBuilder`
</blockquote></details>

I'm worried that there could be concurrency via re-entrancy: when the replacer is constructing new expressions, we might end up building new groups which could require new logical props which could call back into statistics builder which could call this function again.
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 5023 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120875#-NteT9myDNZxUOa2bRu3:-Nts95cm1dx62JAvATv2:b59thqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7531502631fe5f07575492318187aca8365ab53/pkg/sql/opt/memo/statistics_builder.go#L5023)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Note: I believe this is relying on the expression de-duplication performed in the optbuilder to work when the user explicitly writes the expression:
```
https://github.com/cockroachdb/cockroach/blob/4a9385cacb82e7a8d6d37e5d9a26a6b7c845aab6/pkg/sql/opt/optbuilder/project.go#L278-L285
```
I think it wouldn't work if you pruned the `b` column before the filter:
```
SELECT * FROM (SELECT a FROM abc) WHERE a % 10 > 3
```
It should also fail if the expression was copied at any point, which could motivate testing with prepared statements and routines. Prepared statements in particular seem like they could be a problematic omission, since they're so common.
</blockquote></details>

Could you say more about how this is relying on that de-duplication in finishBuildScalar? I thought this was relying on the fact that we memoize scalar expressions whenever we construct them, but maybe I am mistaken.

I added that example as a testcase.

I also added some assign-placeholder-norm testcases. Not sure if that's exactly equivalent to prepared statements, but it does detach the memo at least.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for adding the tests!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120875#-:-NtsF4GtBpgf2g8n3Xoi:b-5edq8h)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 5023 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120875#-NteT9myDNZxUOa2bRu3:-NtsEgDY5dH2agST7rpt:b-1cboif) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7531502631fe5f07575492318187aca8365ab53/pkg/sql/opt/memo/statistics_builder.go#L5023)):*
>I thought this was relying on the fact that we memoize scalar expressions whenever we construct them

Ah, that was the missing piece for me. It's great that it just works out like that!


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Of course. Thank you for all of the reviews!

Failure is https://github.com/cockroachdb/cockroach/issues/121087 which looks unrelated.

bors r=DrewKimball

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120875#-:-Ntt4O7O7cO3hOMveKeX:by2d1wo)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)



<!-- Sent from Reviewable.io -->

michae2 commented:
bors r=DrewKimball
